### ISSUE TITLE
Sorted map please!

### ISSUE THREAD

**Issue by chenglou** [facebook/immutable-js#88]
A map that inserts new items in the correct position, providing a custom sort function on the keys.

Would be nice. Otherwise, I can do fine with keeping a key array for the sorting, and a map myself.

**Comment by Dashed** [comment]
Just figured out I need this.

+1 from me.

**Comment by trevorbernard** [comment]
+1 indeed :shipit:

**Comment by zxqfox** [comment]
:+1:

**Comment by facebook-github-bot-6** [comment]
Thank you for reporting this issue and appreciate your patience. We've  notified the core team for an update on this issue. We're looking for a response within the next 30 days or the issue may be closed.

**Comment by manuel** [comment]
Sorted map would be great! :+1:

**Comment by matklad** [comment]
Also, an `OrderedMap` can be implemented atop of a `SortedMap` without amortization.

**Comment by dozoisch** [comment]
That would be so useful!

**Comment by rhys-vdw** [comment]
Agreed. If I understand correctly `.sort()` creates a copy every time (even when called in `withMutations`). So if I want to insert sorted, I have to duplicate the `OrderedMap` twice. eg:

```js
const map = OrdererMap({a: 'a', c: 'c'});
const sorted = map.set('b', 'b').sort(); // Creates two copies.
```

**Comment by thomasboyt** [comment]
For prior art on this issue, Clojure(script)'s sorted-map implementation uses a persistent red-black tree:

* Java: https://github.com/clojure/clojure/blob/010864f8ed828f8d261807b7345f1a539c5b20df/src/jvm/clojure/lang/PersistentTreeMap.java
* Clojure (used in Clojurescript): https://github.com/clojure/clojurescript/blob/609400910e983849e5b7932a4b7d0b0dda465df6/src/main/cljs/cljs/core.cljs#L7066

There's a JavaScript implementation of the same data structure [here](https://github.com/mikolalysenko/functional-red-black-tree) that might be useful, too.

Clojure's sorted-set uses the same data structure, slightly wrapped: https://github.com/clojure/clojure/blob/clojure-1.7.0/src/jvm/clojure/lang/PersistentTreeSet.java

**Comment by leebyron** [comment]
RB Tree sounds like the right way to do this.

I'm putting this up for grabs for any adventurous contributors. Best way to learn about Persistent Immutable data structures is to implement them!

**Comment by rongierlach** [comment]
Have not used yet but this seems promising: https://github.com/frptools/collectable/tree/master/packages/sorted-map#collectablejs-immutable-sorted-map

**Comment by crodriguez1a** [comment]
@akagomez Want to take a crack at RB Tree implementation?

**Comment by akagomez** [comment]
Tempting @crodriguez1a. 

I've done a bunch of work with Red-Black Binary Tree's for CanJS (https://github.com/canjs/can-binarytree), based on a fork of this repo: https://github.com/vadimg/js_bintrees

Leveraging that RB Tree implementation again should make this a less daunting task. 

If nobody disagrees with this approach I don't mind taking a crack at it with @crodriguez1a's help.

**Comment by rongierlach** [comment]
@akagomez feel free to steal from my toy SortedMap (api not fully implemented) that I started for a side project: https://github.com/rongierlach/immutable-sorted-map/

It is using a regular BST under the hood, am intending to convert it to RB later on.

**Comment by applitopia** [comment]
The long wait is over. Have a look at our immutable-js fork with highly optimized implementation of **SortedMap** and **SortedSet**. Current version is using classic B-trees. https://github.com/applitopia/immutable-sorted

**Comment by akagomez** [comment]
Darn, I was looking forward to playing with this myself. 

Maybe @crodriguez1a and I can help convert @applitopia's implementation to use RB-trees.

**Comment by leebyron** [comment]
@applitopia - please consider contributing those implementations to this project!

**Comment by axefrog** [comment]
For reference if you want to adapt it, or as an alternative solution, my own immutable collections library has a sorted map, a sorted set _and_ a red-black tree. The sorted map/set collections behave like Immutable.js' ordered equivalents by default, or you can supply a custom sorting predicate to have them sort their elements however you like. https://github.com/frptools/collectable

**Comment by jdizzle217** [comment]
Need this too! Thanks
